pkg/stream/producer_daemon.go

```diff
@@ -30,96 +31,104 @@ var (
+		// memory (depending on the number of buckets you configure).
+		//
+		// Note: This still does not guarantee that your messages are perfectly ordered - this is impossible as soon as you
+		// have more than once producer. However, messages with the same partition key will end up in the same shard, so if
+		// you are reading two different shards and one is much further behind than the other, you will not see messages
+		// *massively* out of order - it should be roughly bounded by the time you buffer messages (the Interval setting) and
+		// thus be not much more than a minute (using the default setting) instead of hours (if one shard is half a day behind
+		// while the other is up-to-date).
+		//
+		// Second note: If you change the amount of partitions, messages might move between buckets and thus end up in different
+		// shards than before. Thus, only do this if you can handle it (e.g., because no shard is currently lagging behind).
+		PartitionBucketCount int `cfg:"partition_bucket_count" default:"128" validate:"min=1"`
+
+		// Additional attributes we append to each message
+		MessageAttributes map[string]string `cfg:"message_attributes"`
+	}
 
-type producerDaemon struct {
-	kernel.EssentialBackgroundModule
-
-	name       string
-	lck        sync.Mutex
-	logger     log.Logger
-	metric     metric.Writer
-	aggregator ProducerDaemonAggregator
-	batcher    ProducerDaemonBatcher
-	outCh      OutputChannel
-	output     Output
-	clock      clock.Clock
-	ticker     clock.Ticker
-	settings   ProducerDaemonSettings
-	stopped    int32
-}
+	producerDaemon struct {
+		kernel.EssentialBackgroundModule
+
+		name       string
+		lck        sync.Mutex
+		logger     log.Logger
+		metric     metric.Writer
+		aggregator ProducerDaemonAggregator
+		batcher    ProducerDaemonBatcher
+		outCh      OutputChannel
+		output     Output
+		clock      clock.Clock
+		ticker     clock.Ticker
+		settings   ProducerDaemonSettings
+		stopped    int32
+	}
 
-func ResetProducerDaemons() {
-	producerDaemonLock.Lock()
-	defer producerDaemonLock.Unlock()
+	producerDaemonKey string
+)
 
-	producerDaemons = map[string]*producerDaemon{}
+func producerDaemonName(name string) string {
+	return fmt.Sprintf("producer-daemon-%s", name)
 }
 
 func ProvideProducerDaemon(ctx context.Context, config cfg.Config, logger log.Logger, name string) (*producerDaemon, error) {
-	producerDaemonLock.Lock()
-	defer producerDaemonLock.Unlock()
-
-	if _, ok := producerDaemons[name]; ok {
-		return producerDaemons[name], nil
+	producer, err := appctx.Provide(ctx, producerDaemonKey(producerDaemonName(name)), func() (*producerDaemon, error) {
+		return NewProducerDaemon(ctx, config, logger, name)
+	})
+	if err != nil {
+		return nil, fmt.Errorf("retrieving daemon from appctx: %w", err)
 	}
 
-	var err error
-	producerDaemons[name], err = NewProducerDaemon(ctx, config, logger, name)
-	if err != nil {
-		return nil, err
+	if producer.IsStopped() {
```

Comment:
I would not make IsStopped and Restart public - user code shouldn't need to query them or reset them, if they have to, gosoline itself needs to be improved to cover this use-case